crypt.go

```diff
@@ -0,0 +1,494 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package secboot
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"golang.org/x/sys/unix"
+	"golang.org/x/xerrors"
+)
+
+const userKeyring = -4
+
+var (
+	runDir                = "/run"
+	systemdCryptsetupPath = "/lib/systemd/systemd-cryptsetup"
+)
+
+type execError struct {
+	path string
+	err  error
+}
+
+func (e *execError) Error() string {
+	return fmt.Sprintf("%s failed: %s", e.path, e.err)
+}
+
+func (e *execError) Unwrap() error {
+	return e.err
+}
+
+func wrapExecError(cmd *exec.Cmd, err error) error {
+	if err == nil {
+		return nil
+	}
+	return &execError{path: cmd.Path, err: err}
+}
+
+func isExecError(err error, path string) bool {
+	var e *execError
+	return xerrors.As(err, &e) && e.path == path
+}
+
+func activate(volumeName, sourceDevicePath string, key []byte, options []string) error {
+	keyFilePath, err := func() (string, error) {
```

Comment:
This can be done in a follow-up, but I think it could be interesting to consider using a named pipe or similar mechanism to avoid creating temporary files containing the key.